package/MDAnalysis/core/selection.py

```diff
@@ -678,7 +679,9 @@ def _apply(self, group):
         if not pattern:
             raise ValueError(f"{self.pattern!r} is not a valid SMARTS query")
         mol = group.convert_to("RDKIT", **self.rdkit_kwargs)
-        matches = mol.GetSubstructMatches(pattern, useChirality=True)
+        self.smarts_kwargs.setdefault("useChirality", True)
+        self.smarts_kwargs.setdefault("maxMatches", 1000)
```

Comment:
So re-reading the various comments I actually quite like your suggestion (https://github.com/MDAnalysis/mdanalysis/pull/3470#discussion_r766880225) to use `max(1000, n_atoms)` as the default, I would happy with that being implemented here and skipping the warning / deprecation with the excuse that it's technically a "bug" in large systems.

Reply:
Do we have a sense for how frequently this happens at n_atoms vs 10 * n_atoms, and how large a cost each order of magnitude adds? Is there a way this could easily be benchmarked?

If we think we'll still get a lot of problem cases unless we have to go to a number that is prohibitively expensive, we might want to inspect the return number of matches and raise a warning if maxMatch == len(matches)

Reply:
I performed a quick benchmarking and the match number seems to have a negligible impact on the overall time calling `select_atoms`. The system I explored has ~8000 atoms

```
u.select_atoms(f'smarts C', smarts_kwargs={'maxMatches': 10})
```
and 
```
u.select_atoms(f'smarts C', smarts_kwargs={'maxMatches': 100000})
```
the former returned 10 atoms in 718 ms and the latter returned 2035 atoms in 728 ms, which is correct. It seems that the python code is much slower than the substructure matching written in C. Except for very large systems, I doubt the substructure matching will be an issue.

Reply:
I'm not sure how frequently `n_atoms` is insufficient. I wasn't able to reproduce the issue on the system I have on hand, but I have encountered it in the past.

Reply:
Ok yeah that makes sense it'll just be a loop that terminates once there are no further matches so you won't get any difference in performance unless `maxMatches < all possible matches`.

I _think_ my 2 cent here is that we do the following:

1. We set `maxMatches` to some arbitrary high value if we think that's beneficial in most cases (10 * n_atoms if you want for now).
2. We put up a warning in the docs detailing a) the issue, explaining what happens when you run out of allowable matches, b) how to fix it, c) the fact that this is currently under refinement and the maximum number of matches may grow / reduce in future versions.
3. Issue a warning if the number of matches == maxMatches that tells users that a maximum number of matches was returned and they may need to increase the number of matches if they want. (@jbarnoud @lilyminium I'm terrible at adding extra warnings - you both have stronger views on these things, are you for/against this?)

Reply:
That all sounds good to me. A warning seems wise because a) `matches == maxMatches` almost certainly is not the intended behavior, and b) otherwise it would be a silent error that would be tricky to track down.